### PR TITLE
Feature/integrate amount

### PR DESCRIPTION
**Pull request by kamilsa**
## What is this pull request?
Integrate new amount class into iroha
   
## Why do you implement it? Why do we need this pull request?
New amount stores internally uint256 type, so to integrate this type protobuf and json converters need to be updated.
  
## How to use the features provided in the pull request?
Use new amount class when deal with values of assets. See ametsuchi, converters and amount tests for example

## Details/Features
To send amount via GRPC, we need to represent it in protobuf via for uint64 variables for uint256 value and single precision. See Amount message in primitive.proto for details. 
In postgres new amount is stored as decimal type. See diff in postgres_wsv_query.cpp for reference

### CODE REVIEW

File: irohad/model/converters/impl/json_command_factory.cpp [hunk 1]
```diff
@@ -36,11 +37,26 @@ namespace iroha {
...
+            return nonstd::nullopt;
+          }
+
+          // check if value is actually number
+          std::regex e("\\d+");
+          if (!std::regex_match(str_int_value.value(), e)){
+            return nonstd::nullopt;
+          }
+
+          uint256_t value(str_int_value.value());
```
Comment (x3medima17) [inline irohad/model/converters/impl/json_command_factory.cpp]: Check if str_int_value fits in uint256_t

File: irohad/model/execution/impl/command_executor.cpp [hunk 2]
```diff
@@ -60,7 +60,7 @@ bool AddAssetQuantityExecutor::execute(const Command &command,
   }
   auto precision = asset.value().precision;
 
-  if (add_asset_quantity.amount.get_frac_number() > precision) {
+  if (add_asset_quantity.amount.getPrecision() != precision) {
     log_->info("amount is wrong formed:");
```
Comment (x3medima17) [inline irohad/model/execution/impl/command_executor.cpp]: I think "wrongly" fits better, or malformed

File: irohad/model/execution/impl/command_executor.cpp [hunk 3]
```diff
@@ -71,22 +71,22 @@ bool AddAssetQuantityExecutor::execute(const Command &command,
...
+    account_asset->balance = add_asset_quantity.amount;
   } else {
-    account_asset = account_asset.value();
-    // TODO: handle non trivial arithmetic
-    auto new_balance = account_asset.value().balance +
-        add_asset_quantity.amount.get_joint_amount(precision);
-    // TODO: handle overflow
-    account_asset->balance = new_balance;
+    auto account_asset_value = account_asset.value();
+    std::cout << account_asset_value.balance.to_string() << " " << add_asset_quantity.amount.to_string() << std::endl;
```
Comment (x3medima17) [inline irohad/model/execution/impl/command_executor.cpp]: Do we need std::cout here? Maybe it is better to use Log

File: irohad/model/execution/impl/command_executor.cpp [hunk 4]
```diff
@@ -71,22 +71,22 @@ bool AddAssetQuantityExecutor::execute(const Command &command,
...
+    auto new_balance =
+        account_asset_value.balance + add_asset_quantity.amount;
+    if (!new_balance.has_value()) {
```
Comment (x3medima17) [inline irohad/model/execution/impl/command_executor.cpp]: not vs ! operator

File: irohad/model/execution/impl/command_executor.cpp [hunk 5]
```diff
@@ -439,24 +431,25 @@ bool TransferAssetExecutor::execute(const Command &command,
...
+  if (transfer_asset.amount.getPrecision() != precision) {
     log_->info("precision {} is wrong", precision);
     return false;
   }
   // Get src balance
   auto src_balance = src_account_asset.value().balance;
   // TODO: handle non-trivial arithmetic
-  src_balance -= transfer_asset.amount.get_joint_amount(precision);
+  auto new_src_balance = src_balance - transfer_asset.amount;
+  if (!new_src_balance.has_value()) {
```
Comment (x3medima17) [inline irohad/model/execution/impl/command_executor.cpp]: not vs ! operator

File: irohad/model/execution/impl/command_executor.cpp [hunk 6]
```diff
@@ -466,22 +459,25 @@ bool TransferAssetExecutor::execute(const Command &command,
...
 
   } else {
     // Account already has such asset
     dest_AccountAsset = dest_account_asset.value();
     // Get balance dest account
     auto dest_balance = dest_account_asset.value().balance;
 
-    dest_balance += transfer_asset.amount.get_joint_amount(precision);
+    auto new_dest_balance = dest_balance + transfer_asset.amount;
+    if (!new_dest_balance.has_value()) {
```
Comment (x3medima17) [inline irohad/model/execution/impl/command_executor.cpp]: not vs ! operator

File: irohad/model/execution/impl/command_executor.cpp [hunk 7]
```diff
@@ -492,15 +488,14 @@ bool TransferAssetExecutor::hasPermissions(const Command &command,
...
+         creator.account_id == transfer_asset.src_account_id;
 }
 
 bool TransferAssetExecutor::isValid(const Command &command,
                                     ametsuchi::WsvQuery &queries) {
   auto transfer_asset = static_cast<const TransferAsset &>(command);
 
-  if (not(transfer_asset.amount.frac_part > 0 or
-      transfer_asset.amount.int_part > 0)) {
+  if (not(transfer_asset.amount.getIntValue() > 0)) {
```
Comment (x3medima17) [inline irohad/model/execution/impl/command_executor.cpp]: Why not <= 0 ?

File: libs/amount/amount.cpp [hunk 8]
```diff
@@ -33,13 +34,36 @@ namespace iroha {
...
+    res <<= 64;
+    res |= second;
+    res <<= 64;
+    res |= third;
+    res <<= 64;
+    res |= fourth;
+    return res;
+  }
+
   Amount::Amount() : value_(0), precision_(0) {}
```
Comment (x3medima17) [inline libs/amount/amount.cpp]: Why not setting default value in header file?

File: libs/amount/amount.cpp [hunk 9]
```diff
@@ -33,13 +34,36 @@ namespace iroha {
...
 
   Amount::Amount(uint256_t value) : value_(value), precision_(0) {}
 
   Amount::Amount(uint256_t amount, uint8_t precision)
       : value_(amount), precision_(precision) {}
 
+  Amount::Amount(uint64_t first, uint64_t second, uint64_t third,
```
Comment (x3medima17) [inline libs/amount/amount.cpp]: You can use constructor delegation here

File: irohad/model/execution/impl/command_executor.cpp (hunk 2 above)
Comment (kamilsa) [reply irohad/model/execution/impl/command_executor.cpp]: Not mine code, but ok

File: irohad/model/execution/impl/command_executor.cpp (hunk 3 above)
Comment (kamilsa) [reply irohad/model/execution/impl/command_executor.cpp]: It is better to remove it

File: irohad/model/converters/impl/pb_common.cpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  namespace model {
+    namespace converters {
+
+      protocol::Amount serializeAmount(
+          iroha::Amount iroha_amount) {
+        protocol::Amount res;
+        res.set_precision(iroha_amount.getPrecision());
+        auto value = res.mutable_value();
+        auto vectorUint64s = iroha_amount.to_uint64s();
+        value->set_first(vectorUint64s.at(0));
```
Comment (Warchant) [inline irohad/model/converters/impl/pb_common.cpp]: at throws

File: libs/amount/amount.cpp (hunk 8 above)
Comment (Warchant) [reply libs/amount/amount.cpp]: even better to write in header:
```
uint64_t value_{0};
```

**Pull request merged**